src/utils/references.ts

```diff
@@ -17,136 +17,208 @@
-const replaceAllReferences = (obj: any, references: any, discovered: Map<string, any>, visited: Set<string>): any => {
-	if (isTerminalSymbol(obj)) {
-		// The object could be a terminal symbol already, like a boolean, or a number. In that case, further search is not required.
-		return obj;
+const isReferenceTag = (input: unknown): input is ReferenceTag => ReferenceTag.safeParse(input).success;
+const referenceTagToObjectPath = (value: ReferenceTag): string => {
+	try {
+		return value.split('/').slice(1).join('.');
+	} catch (e) {
+		throw new SyntaxError(`Unable to parse value '${value}' as a ReferenceTag.`);
 	}
-	if (isReferenceValue(obj)) {
-		// If the object represents a reference, then try to resolve it, either by a direct look up or recursively.
-		return resolveReference(obj, references, discovered, visited);
-	} else if (Array.isArray(obj)) {
-		// If the object is an array, each element could potentially be a reference that ends up being resolved to one or more elements.
-		return replaceReferencesInArray(obj, references, discovered, visited);
-	} else {
-		// `obj` is an object, so each property is replaced with the resolved references
-		return replaceReferencesInObject(obj, references, discovered, visited);
-	}
-}
-
-const referenceToObjectPath = (value: string) => {
-	return value
-		.split('/')
-		.slice(1)
-		.join('.');
 };
 
-const isReferenceValue = (value: string) => {
-	const regex = new RegExp('^#(/[-_A-Za-z0-9]+)+$');
-	return regex.test(value);
+const resolveAllReferences = (
+	value: undefined | string | string[],
+	references: References,
+	discovered: DiscoveredMap,
+	visited: VisitedSet,
+): string | string[] | undefined => {
+	if (isArray(value)) {
+		return value.flatMap((item) =>
+			isReferenceTag(item) ? resolveReference(item, references, discovered, visited) : item,
+		);
+	}
+	if (isReferenceTag(value)) {
+		return resolveReference(value, references, discovered, visited);
+	} else {
+		return value;
+	}
 };
 
-// Checks if the symbol is already something that can be replaced in a reference
-const isTerminalSymbol = (value: any): boolean => {
-	return typeof value !== 'object' && !(isReferenceValue(value))
-}
-
-const resolveReference = (reference: string, references: any, discovered: Map<string, any>, visited: Set<string>): any => {
-	if (discovered.has(reference)) {
-		return discovered.get(reference)
+/**
+ * Recursive handler for getting value associated with a reference tag.
+ * This will throw an error if the tag provided finds no matching references, so ensure that the value provided
+ * to the tag argument is a valid ReferenceTag.
+ *
+ * @param tag
+ * @param references
+ * @param discovered
+ * @param visited
+ * @returns
+ */
+const resolveReference = (
+	tag: ReferenceTag,
+	references: References,
+	discovered: DiscoveredMap,
+	visited: VisitedSet,
+): string | string[] => {
+	const cachedValue = discovered.get(tag);
+	if (cachedValue !== undefined) {
```

Comment:
le `!isUndefined` is le blergh

Reply:
For future reference: The intention here is to permit empty strings through, and certainly TS tells us it will only be a string, string array, or undefined... but assuming some strange value gets through this will also permit NaN and null through.

The absolute best check would be to check if it is truthy or an empty string: `if (cachedValue || cachedValue === '')`